src/doppler-types.ts

```diff
@@ -64,39 +64,210 @@ export interface PrepaidPack {
 
 export interface StandardPath {
   type: 'standard';
   actual: boolean;
   minimumFee: number;
-  deadend: boolean;
+  deadEnd: boolean;
 }
 
 export interface PlusPath {
   type: 'plus';
   actual: boolean;
   minimumFee: number;
-  deadend: boolean;
+  deadEnd: boolean;
 }
 
 export interface AgenciesPath {
   type: 'agencies';
   actual: boolean;
-  deadend: true;
+  deadEnd: boolean;
 }
+
+const getPrepaidPacks = (planList: Plan[]): PrepaidPack[] =>
+  planList.filter((x) => x.type == 'prepaid') as PrepaidPack[];
+
+const getUpgradeMonthlyPlans = (
+  planList: Plan[],
+  { minFee, minEmailsByMonth }: { minFee: number; minEmailsByMonth: number } = {
+    minFee: 0,
+    minEmailsByMonth: 0,
+  },
+): MonthlyRenewalDeliveriesPlan[] =>
+  planList.filter(
+    (x) => x.type == 'monthly-deliveries' && x.fee > minFee && x.emailsByMonth >= minEmailsByMonth,
+  ) as MonthlyRenewalDeliveriesPlan[];
+
+const getUpgradeSubscribersPlans = (
+  planList: Plan[],
+  { minFee, minSubscriberLimit }: { minFee: number; minSubscriberLimit: number } = {
+    minFee: 0,
+    minSubscriberLimit: 0,
+  },
+): SubscribersLimitedPlan[] =>
+  planList.filter(
+    (x) => x.type == 'subscribers' && x.fee > minFee && x.subscriberLimit >= minSubscriberLimit,
+  ) as SubscribersLimitedPlan[];
+
+const getFreePlans = (planList: Plan[]): FreePlan[] =>
+  planList.filter((x) => x.type == 'free') as FreePlan[];
+
+const getAgencyPlans = (planList: Plan[]): AgencyPlan[] =>
+  planList.filter((x) => x.type == 'agency') as AgencyPlan[];
+
+const getStandardPlans = (
+  planList: Plan[],
+): (MonthlyRenewalDeliveriesPlan | SubscribersLimitedPlan)[] =>
+  planList.filter((x) => x.featureSet == 'standard') as (
+    | MonthlyRenewalDeliveriesPlan
+    | SubscribersLimitedPlan
+  )[];
+
+const getPlusPlans = (
+  planList: Plan[],
+): (MonthlyRenewalDeliveriesPlan | SubscribersLimitedPlan)[] =>
+  planList.filter((x) => x.featureSet == 'plus') as (
+    | MonthlyRenewalDeliveriesPlan
+    | SubscribersLimitedPlan
+  )[];
+
+const getFreePathOrEmpty = (userPlan: Plan, planList: Plan[]): [] | [FreePath] => {
+  var plans = getFreePlans(planList);
+
+  if (plans.length == 0) {
+    return [];
+  }
+
+  var cheapestPlan = plans[0];
+
+  return [
+    {
+      type: 'free',
+      actual: userPlan.featureSet == 'free',
+      deadEnd: plans.length == 1 && userPlan == cheapestPlan,
+    },
+  ];
+};
+
+const getAgencyPathOrEmpty = (userPlan: Plan, planList: Plan[]): [] | [AgenciesPath] => {
+  var plans = getAgencyPlans(planList);
+
+  if (plans.length == 0) {
+    return [];
+  }
+
+  var cheapestPlan = plans[0];
+
+  return [
+    {
+      type: 'agencies',
+      actual: userPlan.featureSet == 'agency',
+      deadEnd: plans.length == 1 && userPlan == cheapestPlan,
+    },
+  ];
+};
+
+const getStandardPathOrEmpty = (userPlan: Plan, planList: Plan[]): [] | [StandardPath] => {
+  var plans = getStandardPlans(planList).sort((x) => x.fee);
```

Comment:
Here if we have that the userPlan is plus, but we have a standard plan with more subscribers, here would return a path with actual false. 
If the plan is plus we don´t want any standard plan path.

Reply:
You are right, but it should be filtered before, `potentialUpgradePlans` should be filtered based on your criteria, for example:

```diff
  const potentialUpgradePlans =
    userPlan.type == 'free'
      ? planList
      : userPlan.type == 'prepaid'
      ? [
          ...getPrepaidPacks(planList),
          ...getUpgradeMonthlyPlans(planList),
          ...getUpgradeSubscribersPlans(planList),
        ]
      : userPlan.type == 'monthly-deliveries'
      ? getUpgradeMonthlyPlans(planList, {
          minFee: userPlan.fee,
          minEmailsByMonth: userPlan.emailsByMonth,
        })
      : userPlan.type == 'subscribers'
      ? getUpgradeSubscribersPlans(planList, {
          minFee: userPlan.fee,
          minSubscriberLimit: userPlan.subscriberLimit,
        })
      : [];

+  const potentialUpgradePlansFilteredByPlus = userPlan.featureSet == 'plus' 
+    ? getPlusPlans(potentialUpgradePlans) 
+    : potentialUpgradePlans;

-  const potentialAndCurrentPlans = [userPlan, ...potentialUpgradePlans, _agencyPlan]
+  const potentialAndCurrentPlans = [userPlan, ...potentialUpgradePlansFilteredByPlus, _agencyPlan]
```